File: framework/languages/node/e2e/ResilienceLifecycle/Client/Scenarios/rl-a5-provider-flapping-scenario.ts
// RL-A5: Provider lifecycle을 반복해도 current target에 수렴한다 시나리오를 검증한다.
import type { ClientOptions } from '../Support/client-options';
import { postJson } from '../../../http-client';
import { startProvider } from '../Support/managed-provider';
import {
  profileReq,
  waitForProviderTraffic,
  waitPeerPresent,
  waitUntilAvailable,
  waitUntilDown
} from '../Support/resilience-helpers';
import { ensure } from '../Support/scenario-assert';
import type { ScenarioState } from '../Support/scenario-state';
import type { ProfileRes } from '../../Shared/messages';

export async function runRlA5(options: ClientOptions, state: ScenarioState): Promise<void> {
  for (let cycle = 0; cycle < 3; cycle += 1) {
    await stopProviderB(options, state);
    state.providerBProcess = undefined;
    await waitForSurvivingProvider(options.consumerUrl, cycle);

    for (let i = 0; i < 4; i += 1) {
      const marker = `rl-a5-down-${cycle}-${i}`;
      const reply = await postJson<ProfileRes>(options.consumerUrl, '/profile/request', profileReq(marker));
      ensure(reply.providerRid === 'api-a', 'RL-A5 down window did not converge to api-a.');
    }
    await postJson<string[]>(options.providerAUrl, '/evidence/wait', { contains: `marker=rl-a5-down-${cycle}-` });

    const restarted = startProvider({
      providerMain: options.providerMain,
      logDir: options.logDir,
      redisEndpoint: options.redisEndpoint,
      redisKeyPrefix: options.redisKeyPrefix,
      name: `api-b-flap-${cycle}`,
      rid: 'api-b',
      httpUrl: options.providerBUrl,
      channelEndpoint: options.providerBChannelEndpoint,
      evidenceFileName: `api-b-flap-${cycle}.evidence.log`
    });
    await restarted.waitReady();
    state.providerBProcess = restarted;
    await waitUntilAvailable(options.providerBUrl);
    await waitPeerPresent(options.peerLocationUrl, 'api-b');

    await waitForProviderTraffic(options.consumerUrl, `rl-a5-up-${cycle}`, 'api-b');
    await postJson<string[]>(options.providerBUrl, '/evidence/wait', {
      contains: `profile-request|rid=api-b|value=fast|marker=rl-a5-up-${cycle}-`
    });
  }

  console.log('scenario RL-A5 passed');
}

async function stopProviderB(options: ClientOptions, state: ScenarioState): Promise<void> {
  if (state.providerBProcess !== undefined) {
    await state.providerBProcess.stop();
    await waitUntilDown(options.providerBUrl);
    return;
  }
  await postJson(options.providerBUrl, '/shutdown');
  await waitUntilDown(options.providerBUrl);
}

async function waitForSurvivingProvider(consumerUrl: string, cycle: number): Promise<void> {
  const deadline = Date.now() + 15000;
  let consecutiveApiA = 0;
  let attempt = 0;
  while (Date.now() < deadline) {
    const marker = `rl-a5-converge-${cycle}-${attempt++}`;
    const reply = await postJson<ProfileRes>(consumerUrl, '/profile/request', profileReq(marker));
    ensure(reply.value === 'profile:fast', 'RL-A5 convergence probe returned an unexpected value.');
    consecutiveApiA = reply.providerRid === 'api-a' ? consecutiveApiA + 1 : 0;
    if (consecutiveApiA >= 4) {
      return;
    }
    await new Promise((resolve) => setTimeout(resolve, 100));
  }
  throw new Error('RL-A5 down window did not converge to api-a.');
}

File: framework/languages/dotnet/e2e/ResilienceLifecycle/Client/Scenarios/RlA5ProviderFlappingScenario.cs
// Verifies RL-A5 Provider Flapping behavior.
using ResilienceLifecycle.Client.Support;
using ResilienceLifecycle.Shared;
using Zlink.HttpClient;

namespace ResilienceLifecycle.Client.Scenarios;

// RL-A5 verifies provider repeatedly goes down and comes back while traffic converges.
internal static class RlA5ProviderFlappingScenario
{
    public static async Task RunAsync(
        ZLinkHttpClient consumer,
        ZLinkHttpClient registry,
        ResilienceProcessManager processes,
        ZLinkHttpClient providerA,
        ZLinkHttpClient providerB)
    {
        var previousRoutingId = (await registry.Post("/topology/wait")
            .Body(new TopologyWaitReq("api-b", "Ready", 1))
            .Async<TopologyEntryRes[]>()).Body.Single().RoutingId;

        for (var cycle = 0; cycle < 5; cycle++)
        {
            await processes.StopProviderBWithSigtermAsync();

            await registry.Post("/topology/wait")
                .Body(new TopologyWaitReq("api-b", "Ready", 0))
                .Async<TopologyEntryRes[]>();
            await ProviderTrafficProbe.WaitUntilProviderExcludedAsync(
                consumer,
                "api-b",
                $"rl-a5-converge-{cycle}",
                "RL-A5");

            for (var i = 0; i < 10; i++)
            {
                var downMarker = $"rl-a5-down-{cycle}-{i}";
                var downReply = (await consumer.Post("/profile/request")
                    .Body(new ProfileReq("fast", downMarker))
                    .Async<ProfileRes>()).Body;
                ZlinkStreamAssert.Ensure(downReply.ProviderRid == "api-a",
                    "RL-A5 request during the down window did not use api-a.");
            }
            await providerA.Post("/evidence/wait")
                .Body(new EvidenceWaitReq([$"marker=rl-a5-down-{cycle}-9"], []))
                .Async<string[]>();

            var connectionCount = (await consumer.Get("/connections").Async<string[]>()).Body.Length;
            var restarted = await processes.StartProviderBAsync(TimeSpan.FromSeconds(30));
            using var restartedProviderB = ZLinkHttpClient.Create(restarted.Url)
                .Timeout(TimeSpan.FromMinutes(5))
                .Build();
            for (var attempt = 0; attempt < 100; attempt++)
            {
                try
                {
                    var health = await restartedProviderB.Get("/health").AsyncRaw();
                    if (health.Status == 200) break;
                }
                catch
                {
                    // The scenario keeps polling until api-b accepts HTTP traffic again.
                }

                await Task.Delay(100);
            }

            var rows = (await registry.Post("/topology/wait")
                .Body(new TopologyWaitReq("api-b", "Ready", 1))
                .Async<TopologyEntryRes[]>()).Body;
            ZlinkStreamAssert.Ensure(
                rows.Length == 1 && rows[0].Endpoint == restarted.Endpoint
                    && !string.Equals(rows[0].RoutingId, previousRoutingId, StringComparison.Ordinal),
                $"RL-A5 did not converge to exactly the current api-b endpoint "
                + $"(rows={rows.Length}, previous={previousRoutingId}, "
                + $"current={string.Join(',', rows.Select(row => row.Endpoint))}, "
                + $"expected={restarted.Endpoint}, currentRid={rows[0].RoutingId}).");
            previousRoutingId = rows[0].RoutingId;
            await consumer.Post("/connections/wait")
                .Body(new ConnectionWaitReq(
                    ["kind=ConnectionReady", $"remote={restarted.Endpoint}"], connectionCount))
                .Async<string[]>();

            await providerA.Post("/admin/weight/exclude").AsyncRaw();
            await providerA.Post("/admin/weight/wait")
                .Body(new WeightWaitReq(0))
                .AsyncRaw();
            await ProviderTrafficProbe.WaitUntilProviderExcludedAsync(
                consumer,
                "api-a",
                $"rl-a5-directed-{cycle}",
                "RL-A5");

            for (var i = 0; i < 10; i++)
            {
                var reply = (await consumer.Post("/profile/request")
                    .Body(new ProfileReq("fast", $"rl-a5-directed-{cycle}-{i}"))
                    .Async<ProfileRes>()).Body;
                ZlinkStreamAssert.Ensure(
                    reply.ProviderRid == "api-b",
                    "RL-A5 directed request did not use the current provider B.");
            }

            await providerA.Post("/admin/weight/include").AsyncRaw();
            await providerA.Post("/admin/weight/wait")
                .Body(new WeightWaitReq(100))
                .AsyncRaw();
        }

        Console.WriteLine("scenario RL-A5 passed");
    }
}
